### PR TITLE
Implement im2col and update conv2d

### PR DESCRIPTION
**Pull request by philomath213**
## Description
- Implement matrix vertical and horizontal scan.
- Implement im2col (Image Block to Columns) in C++.
- Update conv2d_im2col implementation in order to simplify the API.
- Implement `im2col_encoding`, it will allow user apply a im2col on input matrix and compute the convolution layer using on TenSEAL, there's no need to pad the input and the kernel nor flatting them in Python.
- Update tests.
- Update Convolution notebook.

## Affected Dependencies
None.

## How has this been tested?
- Unit tests.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: tenseal/cpp/tensors/utils/utils.h [hunk 1]
```diff
@@ -10,6 +12,104 @@ namespace tenseal {
...
+template <typename T>
+void horizontal_scan(const vector<vector<T>>& src, vector<T>& dst) {
+    size_t in_height = src.size();
+    size_t in_width = src[0].size();
+
+    dst.resize(in_height * in_width);
+
+    // check if each row size is equals to in_width
+    if (any_of(src.begin(), src.end(),
+               [in_width](vector<T> i) { return i.size() != in_width; })) {
```
Comment (bcebere) [inline tenseal/cpp/tensors/utils/utils.h:27]: const ref

File: tenseal/cpp/tensors/utils/utils.h [hunk 2]
```diff
@@ -10,6 +12,104 @@ namespace tenseal {
...
+template <typename T>
+void vertical_scan(const vector<vector<T>>& src, vector<T>& dst) {
+    size_t in_height = src.size();
+    size_t in_width = src[0].size();
+
+    dst.resize(in_height * in_width);
+
+    // check if each row size is equals to in_width
+    if (any_of(src.begin(), src.end(),
+               [in_width](vector<T> i) { return i.size() != in_width; })) {
```
Comment (bcebere) [inline tenseal/cpp/tensors/utils/utils.h:52]: const ref

File: tenseal/cpp/tensors/utils/utils.h [hunk 3]
```diff
@@ -10,6 +12,104 @@ namespace tenseal {
 using namespace seal;
 using namespace std;
 
+/**
+ * horizontally scan matrix (vector of vectors)
+ **/
+template <typename T>
+void horizontal_scan(const vector<vector<T>>& src, vector<T>& dst) {
```
Comment (bcebere) [inline tenseal/cpp/tensors/utils/utils.h:19]: These operations might be easier to execute using a gsl::multi_span structure. For sure not for this PR, I don't know if they actually help. But they should prevent extra-copies

Comment (philomath213) [reply tenseal/cpp/tensors/utils/utils.h:19]: Yeah, I will check this out, maybe they will be helpful.

File: tenseal/binding.cpp [hunk 4]
```diff
@@ -81,6 +82,20 @@ PYBIND11_MODULE(_tenseal_cpp, m) {
...
+             const size_t stride) {
+              vector<vector<double>> view_as_window;
+              vector<double> final_vector;
+              size_t windows_nb = im2col(input, view_as_window, kernel_n_rows,
+                                         kernel_n_cols, stride);
+              vertical_scan(view_as_window, final_vector);
+              CKKSVector ckks_vector = CKKSVector(ctx, final_vector);
+              return make_pair(ckks_vector, windows_nb);
+          });
+
```
Comment (youben11) [inline tenseal/binding.cpp:98]: Maybe not in this PR, but the encoding shouldn't be doing encryption, just encoding it into a vector, then encryption should be something separate

File: tenseal/cpp/tensors/ckksvector.cpp [hunk 5]
```diff
@@ -564,39 +564,47 @@ CKKSVector& CKKSVector::polyval_inplace(const vector<double>& coefficients) {
...
     if (windows_nb == 0) {
         throw invalid_argument("Windows number can't be zero");
     }
 
-    if (kernel.empty()) {
-        throw invalid_argument("Kernel vector can't be empty");
+    if (kernel.empty() ||
+        (any_of(kernel.begin(), kernel.end(),
+                [](const vector<double>& i) { return i.empty(); }))) {
+        throw invalid_argument("Kernel matrix can't be empty");
```
Comment (youben11) [inline tenseal/cpp/tensors/ckksvector.cpp:583]: What about something like this? to check whether the vector<vector> is a well organized matrix?

```suggestion
    if (kernel.empty() ||
        (any_of(kernel.begin(), kernel.end(),
                [](const vector<double>& i) { return i.size() == kernel[0].size(); }))) {
        throw invalid_argument("Kernel matrix can't be empty");
```

Comment (philomath213) [reply tenseal/cpp/tensors/ckksvector.cpp:583]: The test will be wrong in the case where `kernel[0].size() == 0`.
in other hand `horizontal_scan` always check if the rows have the same size.

**Pull request merged**